[RFC, RFT] realtek: add Kernel 5.10 support

musashino205 opened the pull request:
This patch series adds Kernel 5.10 support to realtek target.

summary:

- platform support: based on current master (5.4)
- RTL8231 support: use exsisting driver (5.4)
- backported drivers from upstream:
  - RTL83xx GPIO controller (gpio-realtek-otto, 5.13)
  - IRQ controller (irq-realtek-rtl, 5.12)
  - SPI controller (spi-realtek-rtl, 5.12)
- add "dts-5.10" directory

Tested on:

- INABA Abaniact AML2-17GP (RTL8382M, 17ports)
- I-O DATA BSH-G24MB (RTL8382M, 24ports, WIP device)
- Panasonic Switch-M8eG PN28080K (RTL8380M, 8ports + 1 SFP, WIP device)

musashino205 committed:
realtek: add Kernel 5.10 as testing version

This patch adds "KERNEL_TESTING_PATCHVER:=5.10" to the Makefile in
realtek target to allow using Kernel 5.10 for testing.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: copy config/files/patches to 5.10

this patch copies the following files from 5.4 to 5.10:

- config-5.4   -> config-5.10
- files-5.4/   -> files-5.10/
- patches-5.4/ -> patches-5.10/

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: refresh patches for Kernel 5.10

This patch refresh all patches in patches-5.10/ to adjust for Kernel
5.10.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: fix "help" line in Kconfig in files/patches for 5.10

In Kernel 5.10, "help" must be used instead of "---help---".

this patch fixes the following errors:

drivers/net/dsa/rtl83xx/Kconfig:7: syntax errorgit
drivers/net/dsa/rtl83xx/Kconfig:6: unknown statement "---help---"
drivers/net/dsa/rtl83xx/Kconfig:7:warning: ignoring unsupported character '.'
drivers/net/dsa/rtl83xx/Kconfig:7: unknown statement "This"
drivers/net/ethernet/Kconfig:170: syntax error
drivers/net/ethernet/Kconfig:169: unknown statement "---help---"
drivers/net/ethernet/Kconfig:170:warning: ignoring unsupported character '.'
drivers/net/ethernet/Kconfig:170: unknown statement "Say"
drivers/net/phy/Kconfig:331: syntax error
drivers/net/phy/Kconfig:330: unknown statement "---help---"
drivers/net/phy/Kconfig:331: unknown statement "Supports"

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: drop fixup_bigphys_addr from ioremap.h in 5.10

A macro with the same name is provided in asm/pgtable.h in Kernel 5.10,
use it and drop from ioremap.h.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: backport gpio-realtek-otto driver from 5.13 to 5.10

This patch backports "gpio-realtek-otto" driver to Kernel 5.10.
"MACH_REALTEK_RTL" is used as a platform name in upstream, but "RTL838X"
is used in OpenWrt, so update the dependency by the additional patch.

Note: GPIO mapping is changed in the upstreamed driver.

  old - new

   24 -  0
   25 -  1
   26 -  2
   27 -  3
   28 -  4
   29 -  5
   30 -  6
   31 -  7
   16 -  8
   17 -  9
   18 - 10
   19 - 11
   20 - 12
   21 - 13
   22 - 14
   23 - 15
    8 - 16
    9 - 17
   10 - 18
   11 - 19
   12 - 20
   13 - 21
   14 - 22
   15 - 23

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: refresh and update config-5.10

This patch adjusts config-5.10 by running kernel_menuconfig.

Note:

- disable psb6970 phy driver (unused in realtek target?)

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: remove unnecessary line from rtl838x/Platform in 5.10

The following line is already defined in arch/mips/Kbuild.platforms by
300-mips-add-rtl838x-platform.patch.

platform-$(CONFIG_RTL838X) += rtl838x/

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: backport spi-realtek-rtl driver from 5.12 to 5.10

This patch backports "spi-realtek-rtl" driver to Kernel 5.10 from 5.12.
"MACH_REALTEK_RTL" is used as a platform name in upstream, but "RTL838X"
is used in OpenWrt, so update the dependency by the additional patch.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: update soc dtsi in 5.10 for backported drivers

this patch updates SoC dtsi (rtl838x.dtsi, rtl930x.dtsi) for the
following backported drivers:

- gpio-realtek-otto (5.13)
- spi-realtek-rtl (5.12)
- irq-realtek-rtl (5.12)

And, disable SoC GPIO node (gpio0) in rtl930x.dtsi in dts-5.10.
Currently, the upstreamed driver doesn't support the GPIO controller on
RTL930x SoC.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: add "soc" node to soc dtsi in dts-5.10

Add a "soc" node as a simple-bus to rtl838x.dtsi and rtl930x.dtsi.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: fix compile errors in rtl838x_eth.c for 5.10

this patch fixes the following errors when compiling:

- "unsigned int txqueue" is added as an additional parameter of
  ndo_tx_timeout in net_device_ops (include/linux/netdevice.h)

- "mac_link_state" in phylink_mac_ops (include/linux/phylink.h)
  is renamed to "mac_pcs_get_state" and changed the return value
  to void from int

- several parameters are added to "mac_link_up" in phylink_mac_ops
  (include/linux/phylink.h) and the order of the parameters is
  changed

  added:
    - int speed
    - int duplex
    - bool tx_pause
    - bool rx_pause

- a parameter "phy_interface_t *interface" is added to of_get_phy_mode
  (drivers/of/of_net.c) and returns the state instead of phy mode

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: fix compile errors in dsa driver for 5.10

this patch fixes the following errors when compiling:

- dsa_switch_alloc is removed[1]

- a parameter "enum dsa_tag_protocol mprot" is added to dsa_tag_protocol
  in dsa_switch_ops (include/net/dsa.h)

- several paramters are added to "phylink_mac_link_up" in dsa_switch_ops
  (include/net/dsa.h)

  added:
    - int speed
    - int duplex
    - bool tx_pause
    - bool rx_pause

- a parameter "struct switchdev_trans *trans" is added to
  port_vlan_filtering in dsa_switch_ops (include/net/dsa.h)

[1]: https://lore.kernel.org/lkml/20191020031941.3805884-17-vivien.didelot@gmail.com/

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: add pinmux node of LED_GLB_CTRL to rtl838x.dtsi in dts-5.10

This patch adds a pinctrl-single pinmux node to allow disabling system
LED and enabling GPIO 0 (old driver: GPIO 24).

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: fix kernel panic in DSA driver for 5.10

dsa_to_port function in 5.10 returns dsa_port from the port list in
dsa_switch_tree, but the tree is built when the switch is registered
by dsa_register_switch and it's null in rtl83xx_mdio_probe.

So, we need to use dsa_to_port after the registration of the switch.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: update GPIO bindings in the dts files in dts-5.10

this patch includes the following changes:

- adjust mapping for the new driver
  - GPIO 24 -> GPIO 0
  - GPIO 47 -> GPIO 0 (+ disabling system LED)

- disable pins in the invalid range
  (out of the range 0-31 of the new driver)
  - are these pins on the external RTL8231 (&gpio1)?
    - GPIO 67 (-> GPIO 3 on &gpio1?)
    - GPIO 94 (-> GPIO 30 on &gpio1?)

- drop "indirect-access-bus-id" property from gpio0 node in device dts
  files

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: backport irq-realtek-rtl driver from 5.12 to 5.10

This patch backports "irq-realtek-rtl" driver to Kernel 5.10 from 5.12.
"MACH_REALTEK_RTL" is used as a platform name in upstream, but "RTL838X"
is used in OpenWrt, so update the dependency by the additional patch.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: drop platform irq driver from 5.10

To use backported irq driver, drop old irq driver from realtek target
and call irqchip_init() in setup.c.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: drop rtl838x spi-nor driver from 5.10

To backport the upstreamed driver (spi-realtek-rtl) from 5.12, drop the
old driver from realtek target.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: drop rtl838x gpio driver from 5.10

To backport the upstreamed driver (gpio-realtek-otto) from 5.13, drop the
old driver from realtek target.

And, modify 301-gpio-add-rtl838x-driver.patch to remove rtl838x GPIO
support and rename it only for rtl8231 GPIO support.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: use gpio-keys instead of "-polled" if SoC GPIO is used in 5.10

The new backported GPIO driver supports interrupt, so use gpio-keys
instead of gpio-keys-polled for keys connected to the internal GPIO
controller.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: use physical addresses in soc dtsi in 5.10

Use physical addresses instead of virtual address in dts files.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: copy dts directory for Kernel 5.10

This patch adds "dts-5.10" directory to use backported drivers.
There are several specification changes in the new drivers, so there
are some compatibility issues in using dts/dtsi files for 5.4.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: cleanup and update soc dtsi in 5.10

the following changes are included in this patch:

- node is enabled by default, drop 'status = "okay"'
- adjust order of "compatible" lines and "reg" lines
- add a new blank line before fixed-link node in rtl830x.dtsi

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 commented on target/linux/realtek/dts-5.10/rtl8382_allnet_all-sg8208m.dts:
```diff
@@ -0,0 +1,149 @@
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
+
+		/* is this pin 3 on the external RTL8231 (&gpio1)? */
+		/*reset {
+			label = "reset";
+			gpios = <&gpio0 67 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};*/
```
Q: ```&gpio0 67``` -> ```&gpio1 3``` ?

musashino205 commented on target/linux/realtek/dts-5.10/rtl8382_d-link_dgs-1210-10p.dts:
```diff
@@ -0,0 +1,155 @@
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
+
+		/* is this pin 30 on the external RTL8231 (&gpio1)? */
+		/*mode {
+			label = "reset";
+			gpios = <&gpio0 94 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};*/
```
Q: ```&gpio0 94``` -> ```&gpio1 30``` ?

dangowrt commented:
Tested on ZyXEL GS1900-10HP (SFP ports untested)
[gs1900hp-boot.txt](https://github.com/openwrt/openwrt/files/6987321/gs1900hp-boot.txt)

musashino205 commented:
tested SFP on Panasonic Switch-M8eG PN28080K (RTL8380M, 8x TP + 1x SFP, WIP device):

- SFP module: 2x Cisco GLC-SX-MM
- opposite device: Panasonic Switch-M24eG PN28240K (stock)

https://gist.github.com/musashino205/3a60d7a024effe13dfe4d688046278ac

bmork commented:
Tested on Netgear GS108Tv3.  Nice work!
[gs108tv3-boot.txt](https://github.com/openwrt/openwrt/files/6990301/gs108tv3-boot.txt)

bkobl commented:
Great work!
I am having problems with the 930x devices, though. The GPIO's dont work. Booting hangs at

`[    0.000000] Linux version 5.10.56 (birger@ThinkPadAMD) (mips-openwrt-linux-musl-gcc (OpenWrt GCC 8.4.0 r17321+31-5c88bf896e) 8.4.0, GNU ld (GNU Binutils) 2.34) #0 Fri Aug 13 15:27:14 2021
[    0.000000] RTL838X model is 0
[    0.000000] RTL839X model is 0
[    0.000000] RTL93XX model is 93021001
[    0.000000] SoC Type: RTL9302B
[    0.000000] Kernel command line: 
[    0.000000] printk: bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019555 (MIPS 34Kc)
[    0.000000] MIPS: machine is Zyxel XGS1210-12 Switch
[    0.000000] Registering _machine_restart
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Using appended Device Tree.
[    0.000000] Primary instruction cache 32kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] On node 0 totalpages: 32768
[    0.000000]   Normal zone: 288 pages used for memmap
[    0.000000]   Normal zone: 0 pages reserved
[    0.000000]   Normal zone: 32768 pages, LIFO batch:7
[    0.000000] pcpu-alloc: s0 r0 d32768 u32768 alloc=1*32768
[    0.000000] pcpu-alloc: [0] 0 
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 32480
[    0.000000] Kernel command line: console=ttyS0,115200
[    0.000000] Dentry cache hash table entries: 16384 (order: 4, 65536 bytes, linear)
[    0.000000] Inode-cache hash table entries: 8192 (order: 3, 32768 bytes, linear)
[    0.000000] Writing ErrCtl register=00060163
[    0.000000] Readback ErrCtl register=00060163
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 115024K/131072K available (5330K kernel code, 608K rwdata, 1180K rodata, 7400K init, 203K bss, 16048K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 256
[    0.000000] random: get_random_bytes called from start_kernel+0x3c4/0x5b4 with crng_init=0
[    0.000000] rtl9300_timer_init: setting up timer
[    0.000000] Frequency in dts: 175000000, my rate is 87500000, period 1750000
[    0.000000] With base b8003200 IRQ: 8
[    0.000000] clocksource: timer: mask: 0xfffffff max_cycles: 0xfffffff, max_idle_ns: 1365186028 ns
[    0.000008] sched_clock: 28 bits at 87MHz, resolution 11ns, wraps every 1533916858ns
[    0.008454] CPU frequency from device tree: 800MHz
[    0.008481] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 4778151116 ns
[    0.023354] sched_clock: 32 bits at 400MHz, resolution 2ns, wraps every 5368709118ns
[    0.031840] Calibrating delay loop... 1064.96 BogoMIPS (lpj=5324800)
[    0.115328] pid_max: default: 32768 minimum: 301
[    0.120499] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.128362] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.138852] dyndbg: Ignore empty _ddebug table in a CONFIG_DYNAMIC_DEBUG_CORE build
[    0.149438] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.160058] futex hash table entries: 256 (order: -1, 3072 bytes, linear)
[    0.167558] pinctrl core: initialized pinctrl subsystem
[    0.173777] NET: Registered protocol family 16
[    0.223219] clocksource: Switched to clocksource MIPS
[    0.229840] NET: Registered protocol family 2
[    0.234661] IP idents hash table entries: 2048 (order: 2, 16384 bytes, linear)
[    0.242986] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.252112] TCP established hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.260363] TCP bind hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.268020] TCP: Hash tables configured (established 1024 bind 1024)
[    0.275097] UDP hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.282135] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.290029] NET: Registered protocol family 1
[    0.432815] workingset: timestamp_bits=14 max_order=15 bucket_order=1
[    0.443793] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.450131] jffs2: version 2.2 (NAND) (SUMMARY) (ZLIB) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
`

Kernel 5.4 would continue like this:
`[    0.490660] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.496912] jffs2: version 2.2 (NAND) (SUMMARY) (ZLIB) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.527500] Probing RTL838X GPIOs for 00009302
[    0.532295] Found RTL9302 GPIO
`
I dont know whether this is a dts problem (am using the one provided) or a problem of the driver. I am not sure that driver was developed for the newer 93xx SoCs.

svanheule commented:
>  am having problems with the 930x devices, though. The GPIO's dont work.

That's to be expected. The port order is reversed compared to rtl83xx. Since I didn't have any 93xx hardware to test, I didn't implement those GPIO peripherals yet.

As a stop-gap solution, I think you should be able to use the fallback compatible with 32 pins, and reversed port order.

musashino205 commented:
Thank you!
I didn't know details about the new driver with RTL93xx SoCs too, but there is a SoC gpio node with ```compatible = "realtek,rtl838x-gpio"``` in rtl930x.dtsi and I guessed the driver can also be used for RTL930x.dtsi. (I don't have any device with RTL930x SoC...)

> That's to be expected. The port order is reversed compared to rtl83xx. Since I didn't have any 93xx hardware to test, I didn't implement those GPIO peripherals yet.

Oh, I see. There is no officially supported RTL930x device in OpenWrt yet, so I would like to temporarily disable or remove the SoC GPIO node in rtl930x.dtsi for 5.10.

bkobl commented:
Disabling the 93xx GPIO support for now is fine for me. It is more important that the backported drivers plus the 5.10 support make it into OpenWRT and we can build up from there.

musashino205 commented:
UPDATE:

- disable SoC GPIO node (```gpio0```) in rtl930x.dtsi (9dbd5278b789d6002edc94d010f6c635d63aa694)
- drop ```indirect-access-bus-id``` property from device dts files in (9ad67df7ff1c98ebd74e27887206c7020e5d578e)

Neustradamus commented:
To follow this PR

Borromini commented:
I just tested this on my GS1900-8HP v1. Seems to works nice, kudos @musashino205 

Question: PoE doesn't seem to turn off between reboots, is that considered normal (or desired) behaviour? I know both my GS1900-10HP and the -8HP v1 would kill PoE clients when rebooting (same with early support in master and now in the 21.02 branch).

I tested this alongside blogic's PoE C implementation, which doesn't seem to pick up anything of the PoE internals . Still have to reference that with the Lua implementation on 5.10 though.

bmork commented:
> Question: PoE doesn't seem to turn off between reboots, is that considered normal (or desired) behaviour?

Interesting!  I'd definitely say it's desired.

musashino205 commented:
Thank you for the testing, I have no device with PoE support and didn't test about it...
The micro-controller for PoE is controlled via secondary UART on SoC? If so, I've disabled it in SoC dtsi files and it need to be enabled in device dts to use.

musashino205 committed:
realtek: enable uart1 on the devices with PoE support in 5.10

On the devices with PoE support, the secondary UART (uart1) on the SoC
is used to communicate between the SoC and controller.

Enable the secondary UART on the following devices:

- D-Link DGS-1210-10P
- Netgear GS110TPP v1
- Netgear GS310TP v1
- ZyXEL GS1900-8HP v1/v2
- ZyXEL GS1900-10HP
- ZyXEL GS1900-24HP v2

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 commented:
UPDATE:

- enable ```uart1``` on the following devices for PoE
  - D-Link DGS-1210-10P
  - Netgear GS110TPP v1
  - Netgear GS310TP v1
  - Zyxel GS1900-8HP v1/v2
  - Zyxel GS1900-10HP

Borromini commented:
@musashino205 Thanks, that makes the PoE daemon output valid info indeed!

```
# ubus call poe info
{
	"firmware": "v17.1",
	"mcu": "ST Micro ST32F100 Microcontroller",
	"budget": 70.000000,
	"consumption": 3.100000,
	"ports": {
		"lan1": {
			"priority": 2,
			"mode": "PoE+",
			"status": "Searching"
		},
		"lan2": {
			"priority": 2,
			"mode": "PoE+",
			"status": "Delivering power",
			"consumption": 3.200000
		}
	}
}
```

あろがとう！

dangowrt reviewed (approved):
Run-tested on ZyXEL GS1900-10HP, uptime of 10 days for now, all good.
I've looked at all the commits and didn't see any issues.

Neustradamus commented:
What is the blocking part before a merging?

musashino205 commented:
UPDATE:

- rebase from master for GS1900-24HP v2
  - enable uart1 for GS1900-24HP v2

adschm commented:
Since this already contains quite a lot of patches already and has received positive feedback, I consider merging it during the weekend so it's easier to go on with it (it's just testing for now).

I've reordered the patches a bit, moved dts to dts-5.4 and am currently doing build-testing ...

adschm commented:
I've merged it. Thanks for bringing everything into a nice shape.
